articles/execute_process_refactor.md

```diff
@@ -0,0 +1,619 @@
+| ----------------------------------------- | ------------------------------------------------------------ |
+| `launch.events.lifecycle.StateTransition` | This event is emitted when a message is published to the "`/<node_name>/transition_event`" topic, indicating the lifecycle node represented by this action changed state |
+
+### launch_ros.traits.ComposesNodes
+
+This class would extend the `launch_ros.traits.NodeTrait` class, and provide the additional information required for launching composable nodes that are currently defined in `launch_ros.descriptions.ComposableNode`. 
+
+#### Constructor
+
+Most parameters would be passed to the new superclass.
+
+|Argument|Description|
+|---|---|
+|nodes|Collection of `launch_ros.descriptions.ComposableNode` objects that should be launched inside the `launch_ros.descriptions.Node` with this trait.|
+
+#### Properties
+
+Accessors would be provided for the constructor parameter: `nodes`.
+
+#### Methods
+
+| Name    | Description                                                  |
+| ------- | ------------------------------------------------------------ |
+| prepare | Overridden. Adds an event handler to launch the defined composable nodes in the container once the container is started. |
+
+#### Events
+
+No events would be handled or emitted.
+
+### launch_ros.traits.IsExecutable
+
+This class would extend the `launch_ros.traits.NodeTrait` class, and provide the additional information required for launching executable nodes that are currently defined in `launch_ros.actions.Node`. 
+
+#### Constructor
+
+Most parameters would be passed to the new superclass.
+
+|Argument|Description|
+|---|---|
+|node_executable|Name of the executable to find|
+
+#### Properties
+
+Accessors would be provided for the constructor parameter: `node_exectuable`.
+
+#### Methods
+
+No custom methods would be defined or overridden.
+
+#### Events
+
+No events would be handled or emitted.
+
+### launch_ros.descriptions.RosExecutable
+
+This class would represent the information required to run a ROS-aware process and inherit from `launch.descriptions.Executable`.
+It would not include execution-time details, nor provide monitoring of the process after launch. 
+
+#### Constructor
+
+| Argument        | Description                                                  |
+| --------------- | ------------------------------------------------------------ |
+| nodes             | A list of `launch_ros.descriptions.Node` objects which are part of the executable. |
+
+Additional parameters that may be passed, which are handled by `launch.actions.ExecuteProcess`: `cmd`, `name`, `cwd`, `env`, `additional_env`.
+NOTE: To allow ROS to determine the appropriate executable based on the package and executable name, the `cmd` parameter should *NOT* be specified. When `cmd` is not specified, this class will determine the appropriate executable if and only if exactly one of the nodes it contains has the `launch_ros.traits.IsExecutable` trait.
```

Comment:
@roger-strain why can't the `IsExecutable` trait be subsumed into this ` Executable` subclass, taking package name and executable name as arguments (besides node descriptions)? The current `Node` equivalent would then be:

```python
ExecuteLocal(
   ROSExecutable(
       package_name='my_package',
       executable_name='my_node_executable',
       nodes=[
           Node(name='my_node')
       ]
   )
)
```

Reply:
I like this. I've made that change (though the parameter names may need to be tweaked). See how it looks.

Reply:
I think we can drop the note entirely. `cmd` is not a valid argument for `RosExecutable`'s constructor. No need to ask for it to not be used. If it comes along, `TypeError: invalid arguments` should be raised.